File: KoKoslan/src/java/ast/KoKoCall.java
package kokoslan.java.ast;
import java.util.*;
import java.io.*;
import kokoslan.java.eval.*;

public class KoKoCall implements KoKoAst{
    protected KoKoList args; //Args is what is inside the parenthesis of a call to kokoslan function
    protected KoKoAst head;  //THE NAME OF THE CALL EXAMPLE THE PRINT

    public KoKoCall(KoKoAst head, KoKoList args){
        this.head = head;
        this.args = args;
    }

    public KoKoCall(KoKoAst head){
        this(head, new KoKoList());
    }

    public void genCode(PrintStream out){ //The code we see when we do call.kl
        this.head.genCode(out);
        out.print(" (");
        this.args.genCode(out);
        out.print(") ");
    }

    public KoKoValue eval(KoKoContext ctx){ //Evaluates the call, returns a value, context = id(args)
        return new KoKoEvaluator(this.head, this.args).evaluate(ctx);
    }
    
}

File: KoKoslan/src/kotlin/ast/KoKoCall.kt
/** 
   Andrey Arguedas Espinoza
   Daniela Armas Sanchez
   Michael Chen Wang
   Kimberly Olivas Delgado
*/

package kokoslan.kt.ast

import java.util.*
import java.io.*
import kokoslan.kt.eval.*
import kokoslan.kt.exception.*

class KoKoCall(protected var head : KoKoAst, protected var args : KoKoList = KoKoList()) : KoKoAst {

    override fun genCode(Out : PrintStream) : Unit { //The code we see when we do call.kl
        this.head.genCode(Out)
        Out.print(" (")
        this.args.genCode(Out)
        Out.print(") ")
    }

    override fun eval(ctx : KoKoContext) : KoKoValue? { //Evaluates the call, returns a value, context = id(args)
        return KoKoEvaluator(this.head, this.args).evaluate(ctx)
    }
    
}